core/silkworm/consensus/consensus_engine.cpp

```diff
@@ -0,0 +1,166 @@
+}
+
+std::optional<intx::uint256> expected_base_fee_per_gas(const BlockHeader& header, const BlockHeader& parent,
+                                                              const ChainConfig& config) {
+    if (config.revision(header.number) < EVMC_LONDON) {
+        return std::nullopt;
+    }
+
+    if (header.number == config.revision_block(EVMC_LONDON)) {
+        return param::kInitialBaseFee;
+    }
+
+    const uint64_t parent_gas_target{parent.gas_limit / param::kElasticityMultiplier};
+
+    assert(parent.base_fee_per_gas.has_value());
+    const intx::uint256 parent_base_fee_per_gas{*parent.base_fee_per_gas};
+
+    if (parent.gas_used == parent_gas_target) {
+        return parent_base_fee_per_gas;
+    }
+
+    if (parent.gas_used > parent_gas_target) {
+        const intx::uint256 gas_used_delta{parent.gas_used - parent_gas_target};
+        intx::uint256 base_fee_per_gas_delta{parent_base_fee_per_gas * gas_used_delta / parent_gas_target /
+                                             param::kBaseFeeMaxChangeDenominator};
+        if (base_fee_per_gas_delta < 1) {
+            base_fee_per_gas_delta = 1;
+        }
+        return parent_base_fee_per_gas + base_fee_per_gas_delta;
+    } else {
+        const intx::uint256 gas_used_delta{parent_gas_target - parent.gas_used};
+        const intx::uint256 base_fee_per_gas_delta{parent_base_fee_per_gas * gas_used_delta / parent_gas_target /
+                                                   param::kBaseFeeMaxChangeDenominator};
+        if (parent_base_fee_per_gas > base_fee_per_gas_delta) {
+            return parent_base_fee_per_gas - base_fee_per_gas_delta;
+        } else {
+            return 0;
+        }
+    }
+}
+
+bool is_kin(const BlockHeader& branch_header, const BlockHeader& mainline_header,
+                   const evmc::bytes32& mainline_hash, unsigned n, const State& state,
+                   std::vector<BlockHeader>& old_ommers) {
+    if (n == 0 || branch_header == mainline_header) {
+        return false;
+    }
+
+    std::optional<BlockBody> mainline_body{state.read_body(mainline_header.number, mainline_hash)};
+    if (!mainline_body) {
+        return false;
+    }
+    old_ommers.insert(old_ommers.end(), mainline_body->ommers.begin(), mainline_body->ommers.end());
+
+    std::optional<BlockHeader> mainline_parent{get_parent(state, mainline_header)};
+    std::optional<BlockHeader> branch_parent{get_parent(state, branch_header)};
+
+    if (!mainline_parent) {
+        return false;
+    }
+
+    bool siblings{branch_parent == mainline_parent};
+    if (siblings) {
+        return true;
+    }
+
+    return is_kin(branch_header, *mainline_parent, mainline_header.parent_hash, n - 1, state, old_ommers);
+}
+
+ConsensusEngine& get_consensus_engine(SealEngineType engine_type) {
+    ConsensusEngine* engine;
+    switch (engine_type) {
+        case SealEngineType::kEthash:
+            engine = new Ethash();
```

Comment:
This code leaks memory: you allocate memory with new, but never delete it. Return `std::unique_ptr< ConsensusEngine>` instead since [unique_ptr](https://en.cppreference.com/w/cpp/memory/unique_ptr) will automatically delete its memory when no longer in use.